File: clients/scala/src/main/scala/spb_ai_champ/model/Resource.scala
package spb_ai_champ.model

import spb_ai_champ.util.StreamUtil

/**
 * Resource type
 */
sealed abstract class Resource (val tag: Int) {
    /**
     * Write Resource to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, tag)
    }
}

object Resource {
    /**
     * Stone
     */
    case object STONE extends Resource(0)
    /**
     * Ore
     */
    case object ORE extends Resource(1)
    /**
     * Sand
     */
    case object SAND extends Resource(2)
    /**
     * Organics
     */
    case object ORGANICS extends Resource(3)
    /**
     * Metal
     */
    case object METAL extends Resource(4)
    /**
     * Silicon
     */
    case object SILICON extends Resource(5)
    /**
     * Plastic
     */
    case object PLASTIC extends Resource(6)
    /**
     * Chip
     */
    case object CHIP extends Resource(7)
    /**
     * Accumulator
     */
    case object ACCUMULATOR extends Resource(8)

    /**
     * Read Resource from input stream
     */
    def readFrom(stream: java.io.InputStream): Resource =
        StreamUtil.readInt(stream) match {
            case STONE.tag => STONE
            case ORE.tag => ORE
            case SAND.tag => SAND
            case ORGANICS.tag => ORGANICS
            case METAL.tag => METAL
            case SILICON.tag => SILICON
            case PLASTIC.tag => PLASTIC
            case CHIP.tag => CHIP
            case ACCUMULATOR.tag => ACCUMULATOR
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
}

File: clients/typescript/src/model/resource.ts
import { Stream } from "../stream";

/**
 * Resource type
 */
export class Resource {
    readonly name: string;
    readonly tag: number;

    constructor(name: string, tag: number) {
        this.name = name;
        this.tag = tag;
    }

    /**
     * Stone
     */
    static readonly STONE = new Resource("STONE", 0);
    /**
     * Ore
     */
    static readonly ORE = new Resource("ORE", 1);
    /**
     * Sand
     */
    static readonly SAND = new Resource("SAND", 2);
    /**
     * Organics
     */
    static readonly ORGANICS = new Resource("ORGANICS", 3);
    /**
     * Metal
     */
    static readonly METAL = new Resource("METAL", 4);
    /**
     * Silicon
     */
    static readonly SILICON = new Resource("SILICON", 5);
    /**
     * Plastic
     */
    static readonly PLASTIC = new Resource("PLASTIC", 6);
    /**
     * Chip
     */
    static readonly CHIP = new Resource("CHIP", 7);
    /**
     * Accumulator
     */
    static readonly ACCUMULATOR = new Resource("ACCUMULATOR", 8);

    /**
     * Read Resource from input stream
     */
    static async readFrom(stream: Stream): Promise<Resource> {
        const tag = await stream.readInt();
        if (tag == Resource.STONE.tag) {
            return Resource.STONE;
        }
        if (tag == Resource.ORE.tag) {
            return Resource.ORE;
        }
        if (tag == Resource.SAND.tag) {
            return Resource.SAND;
        }
        if (tag == Resource.ORGANICS.tag) {
            return Resource.ORGANICS;
        }
        if (tag == Resource.METAL.tag) {
            return Resource.METAL;
        }
        if (tag == Resource.SILICON.tag) {
            return Resource.SILICON;
        }
        if (tag == Resource.PLASTIC.tag) {
            return Resource.PLASTIC;
        }
        if (tag == Resource.CHIP.tag) {
            return Resource.CHIP;
        }
        if (tag == Resource.ACCUMULATOR.tag) {
            return Resource.ACCUMULATOR;
        }
        throw new Error("Unexpected tag value");
    }

    /**
     * Write Resource to output stream
     */
    async writeTo(stream: Stream) {
        await stream.writeInt(this.tag);
    }

    [Symbol.for('nodejs.util.inspect.custom')]() {
        return this.name;
    }
}